File: bio17.R
#bio17.R
#To draw graphs...

pdf("bio17.pdf", family="Japan1") #family modifier is required!

#ExpA
b17a <- read.table(file="bio17a.csv",header=T,sep=",")
Deg1=b17a$Deg[1:7];f1=b17a$f[1:7];F1=b17a$F[1:7];L1=b17a$L[1:7]
Deg2=b17a$Deg[8:14];f2=b17a$f[8:14];F2=b17a$F[8:14];L2=b17a$L[8:14]
Deg3=b17a$Deg[15:20];f3=b17a$f[15:20];F3=b17a$F[15:20];L3=b17a$L[15:20]
layout(matrix(c(1,2,3,4),nrow=2,ncol=2,byrow=T))
plot(c(Deg1,Deg2),c(f1,f2),xlab="肘関節角度(deg)",ylab="力(P/P0)",main="角度-力(f)",col=c(rep("black",7),rep("red",7)))
lines(f3 ~ Deg3,col="black")
plot(c(Deg1,Deg2),c(F1,F2),xlab="肘関節角度(deg)",ylab="力",main="角度-力(F)",col=c(rep("black",7),rep("red",7)))
lines(F3 ~ Deg3,col="black")
plot(c(L1,L2),c(F1,F2),xlab="筋長",ylab="力",main="長さ-力(F)",col=c(rep("black",7),rep("red",7)))
lines(F3 ~ L3,col="black")
#ExpB
#a/P0=0.360, b=539.4 these are from Analysis software; you cannot use them as your data.

b17b <- read.table(file="bio17b.csv",header=T,sep=",")
P1=b17b$P[1:9];V1=b17b$V[1:9];W1=b17b$W[1:9]
P2=b17b$P[10:17];V2=b17b$V[10:17];W2=b17b$W[10:17]

#init
x<-seq(0,1,length=1001)
plot(c(P1,P2),c(V1,V2),xlim=c(0,1),ylim=c(0,1500),xlab="力(P/P0)",ylab="角速度(x100deg/s)",main="力-角速度, 力-仕事率",col=c(rep("black",9),rep("red",8)))
par(new=T)
plot(x,539.4*(1-x)/(x+0.360),xlim=c(0,1),ylim=c(0,1500),type="l",xlab="",ylab="",axes=F)
par(new=T)
plot(c(P1,P2),c(W1,W2),xlim=c(0,1),ylim=c(0,300),xlab="",ylab="",col=c(rep("green",9),rep("blue",8)),axes=F)
axis(4)
par(new=T)
plot(x,539.4*x*(1-x)/(x+0.360),xlim=c(0,1),ylim=c(0,300),type="l",xlab="",ylab="",col="pink",axes=F)
dev.off()

File: bio17.rb
#Analysis
#Preparation for bio17.R
#You are forbidden from using this DATA as your experiment's (though program is free).

#実行方法
#1.irbでload "main.rb"した後にbio17a
#2.r bio17.R

def bio17a
	#data: deg, P, P/P0, 1/sin * P/P0, 1-cos/2
	data1=[[90,353.348],[110,349.793],[130,346.237],[150,266.828],[70,270.976],[50,128.158],[90,323.718]]
	data2=[[90,282.236],[110,277.495],[130,268.606],[150,236.012],[70,252.013],[50,199.863],[90,278.680]]
	data=[[50],[70],[90],[110],[130],[150]]
	m1=0; m2=0; m3=0
	data1.sort! {|a,b| x=a[0]<=>b[0]; if x!=0 then x else a[1]<=>b[1] end}
	data2.sort! {|a,b| x=a[0]<=>b[0]; if x!=0 then x else a[1]<=>b[1] end}
	data1.each{|x| if m1<x[1] then m1=x[1] end
		case x[0]
			when 50 then data[0].push(x[1])
			when 70 then data[1].push(x[1])
			when 90 then data[2].push(x[1])
			when 110 then data[3].push(x[1])
			when 130 then data[4].push(x[1])
			when 150 then data[5].push(x[1])
		end
	}
	data2.each{|x| if m2<x[1] then m2=x[1] end
		case x[0]
			when 50 then data[0].push(x[1])
			when 70 then data[1].push(x[1])
			when 90 then data[2].push(x[1])
			when 110 then data[3].push(x[1])
			when 130 then data[4].push(x[1])
			when 150 then data[5].push(x[1])
		end
	}
	data.length.times{|i| data[i][1]=data[i][1..4].average;p data[i]}
	data.each{|x| if m3<x[1] then m3=x[1] end}

	data1.each{|x| x[2]=x[1]/m1; x[3]=x[2]/Math.sin(degtorad(x[0])); x[4]=1-Math.cos(degtorad(x[0]))}
	data2.each{|x| x[2]=x[1]/m2; x[3]=x[2]/Math.sin(degtorad(x[0])); x[4]=1-Math.cos(degtorad(x[0]))}
	data.each{|x| x[2]=x[1]/m3; x[3]=x[2]/Math.sin(degtorad(x[0])); x[4]=1-Math.cos(degtorad(x[0]))}
	File.open("bio17a.csv","w"){|f|
		f.print "Deg,f,F,L\n"
		data1.each{|x| f.print x[0].to_s+","+x[2].to_s+","+x[3].to_s+","+x[4].to_s+"\n"}
		data2.each{|x| f.print x[0].to_s+","+x[2].to_s+","+x[3].to_s+","+x[4].to_s+"\n"}
		data.each{|x| f.print x[0].to_s+","+x[2].to_s+","+x[3].to_s+","+x[4].to_s+"\n"}
	}
end

def bio17b
	#data: P/P0, V
	data1=[[0.281,461.402],[0.410,636.645],[0.391,248.447],[0.466,279.504],[0.691,130.694],[0.906,62.112],[0.995,20.704],[0.258,745.340],[0.438,590.061]]
	data2=[[0.347,402.783],[0.439,361.116],[0.576,208.333],[0.552,319.444],[0.632,111.111],[0.795,41.667],[0.963,0.591],[0.342,569.443]]
	data1.each{|x| x[2]=x[0]*x[1]}
	data2.each{|x| x[2]=x[0]*x[1]}
	File.open("bio17b.csv","w"){|f|
		f.print "P,V,W\n"
		data1.each{|x| f.print x[0].to_s+","+x[1].to_s+","+x[2].to_s+"\n"}
		data2.each{|x| f.print x[0].to_s+","+x[1].to_s+","+x[2].to_s+"\n"}
	}
end
